File: src/java/rules_intellij/worker/IndexingWorker.java
package rules_intellij.worker;

import java.io.*;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import com.beust.jcommander.JCommander;
import com.beust.jcommander.Parameter;

import com.google.common.util.concurrent.FutureCallback;
import com.google.devtools.build.lib.worker.WorkerProtocol.*;
import com.intellij.indexing.shared.ultimate.persistent.rpc.IndexRequest;
import com.intellij.indexing.shared.ultimate.persistent.rpc.IndexResponse;

public final class IndexingWorker {

    static class WorkerArgs {
        @Parameter(names = "--persistent_worker")
        public boolean isPersistent;

        @Parameter(names = "--debug_log")
        public String debugLog;

        @Parameter(names = "--debug_endpoint")
        public String debugEndpoint;

        @Parameter(names = "--project_dir")
        public String projectDir;

        @Parameter(names = "--out_dir")
        public String outDir;

        @Parameter(names = "--target")
        public String target;

        @Parameter(names = "--name")
        public String name;

        @Parameter(names = "-s")
        public List<String> sources = new ArrayList<>();
    }

    private static void startIntellij() {
        // dump-shared-index persistent-project
        // -Dcaches.indexerThreadsCount=1
        throw new UnsupportedOperationException("not implemented yet");
    }

    private static void onStartup(String[] rawArgs, WorkerArgs args, DebugLogger logger) throws IOException {
        logger.log("ARGS", (out) -> {
            for (String arg: rawArgs) {
                out.println(arg);
            }
        });
        logger.log("ENV", (out) -> {
            for (Map.Entry<String, String> entry : System.getenv().entrySet()) {
                out.println(entry.getKey() + " : " + entry.getValue());
            }
        });

        if (args.debugEndpoint == null) {
            startIntellij();
        }
    }

    private static class Work extends WorkerArgs {
        private final DebugLogger logger;
        private final IntellijIndexingPersistentClient client;
        private final long projectId;

        public Work(DebugLogger logger, IntellijIndexingPersistentClient client, long projectId) {
            this.logger = logger;
            this.client = client;
            this.projectId = projectId;
        }

        void processRequest(WorkRequest workRequest) {
            logger.log("WorkRequest", workRequest);

            IndexRequest request = IndexRequest.newBuilder()
                .setProjectId(projectId)
                .setIndexDebugName(target)
                .setIndexId(target
                    .replace("//", "")
                    .replace("/","_")
                    .replace(":", "_"))
                .addAllSources(sources)
                .setProjectRoot(System.getProperty("user.dir"))
                .build();

            client.index(
                request,
                new FutureCallback<IndexResponse>() {
                    private final Path outPath = Paths.get(outDir);

                    private void moveOrThrow(String input, String postfix) throws IOException {
                        File output = outPath.resolve(name + postfix).toFile();
                        if (new File(input).renameTo(output)) {
                            return;
                        }
                        throw new IOException("Can't move " + input + " to " + output.toString());
                    }

                    @Override
                    public void onSuccess(IndexResponse indexResponse)  {
                        try {
                            moveOrThrow(indexResponse.getIjxPath(), ".ijx");
                            moveOrThrow(indexResponse.getIjxMetadataPath(), ".ijx.metadata.json");
                            moveOrThrow(indexResponse.getIjxSha256Path(), ".ijx.sha256");

                            WorkResponse workResponse = WorkResponse.newBuilder()
                                .setRequestId(workRequest.getRequestId())
                                .build();
                            logger.log("WorkResponse", workResponse);
                            workResponse.writeDelimitedTo(System.out);
                        } catch (Throwable error) {
                            onFailure(error);
                        }
                    }

                    @Override
                    public void onFailure(Throwable error) {
                        try {
                            WorkResponse workResponse = WorkResponse.newBuilder()
                                .setExitCode(1)
                                .setOutput(error.toString())
                                .setRequestId(workRequest.getRequestId())
                                .build();

                            logger.log("WorkResponse", workResponse);
                            workResponse.writeDelimitedTo(System.out);
                        } catch (Throwable ioError) {
                            logger.log("UnrecoverableError", ioError.toString());
                            System.exit(1);
                        }
                    }
                }
            );
        }

    }

    public static void main(String[] args) throws IOException {
        WorkerArgs startupArgs = new WorkerArgs();
        JCommander
            .newBuilder()
            .addObject(startupArgs)
            .build()
            .parse(args);

        if (!startupArgs.isPersistent) {
            throw new UnsupportedOperationException("Only persistent workers supported");
        }

        DebugLogger logger = new DebugLogger(startupArgs.debugLog);
        onStartup(args, startupArgs, logger);

        IntellijIndexingPersistentClient client = new IntellijIndexingPersistentClient(logger, startupArgs.debugEndpoint);
        long projectId = client.start(startupArgs.projectDir);

        while (true) {
            WorkRequest request = WorkRequest.parseDelimitedFrom(System.in);
            if (request == null) {
                break;
            }

            Work worker = new Work(logger, client, projectId);
            JCommander.newBuilder()
                .addObject(worker)
                .build()
                .parse(request.getArgumentsList().toArray(new String[] {}));

            worker.processRequest(request);
        }
    }

}

File: src/main/kotlin/rules_intellij/worker/IndexingWorker.kt
package rules_intellij.worker

import com.google.devtools.build.lib.worker.WorkerProtocol.WorkRequest
import com.google.devtools.build.lib.worker.WorkerProtocol.WorkResponse
import com.intellij.indexing.shared.ultimate.persistent.rpc.IndexRequest
import kotlinx.coroutines.*
import java.io.File
import java.io.IOException
import java.nio.file.Paths

object IndexingWorker {

    @JvmStatic
    fun main(args: Array<String>) = runBlocking {
        val logger = WorkerLogger(System.getenv("INTELLIJ_WORKER_DEBUG"))
        onStartup(args, logger)

        val startupArgs = IndexingWorkerArgs()
            .also { it.parseArgs(args) }

        if (!startupArgs.isPersistent) {
            throw UnsupportedOperationException("Only persistent workers supported")
        }

        try {
            val ioScope = CoroutineScope(Dispatchers.IO)
            val client = createIntellijIndexingClient(startupArgs, logger).getOrThrow()
            client.start().getOrThrow()

            while (true) {
                val request = WorkRequest.parseDelimitedFrom(System.`in`)
                    ?: break

                ioScope.launch {
                    logger.log("WorkRequest", request)
                    try {
                        val projectId = client.getProjectId().getOrThrow()
                        processWorkRequest(client, projectId, request)
                            .also { logger.log("WorkResponse", it) }
                            .apply { synchronized(IndexingWorker) {
                                writeDelimitedTo(System.out)
                            } }

                    } catch (e: Exception) {
                        logger.err("UnrecoverableError", e)
                    }
                }
            }
        } catch (e: Exception) {
            logger.err("UnrecoverableError", e)
        }
    }

    private fun onStartup(rawArgs: Array<String>, logger: WorkerLogger) {
        logger.log("ARGS") {
            for (arg in rawArgs) {
                it.println(arg)
            }
        }
        logger.log("ENV") {
            for ((key, value) in System.getenv()) {
                it.println("$key : $value")
            }
        }
    }

    private suspend fun processWorkRequest(
        client: IntellijIndexingClient,
        projectId: Long,
        workRequest: WorkRequest
    ): WorkResponse {
        try {
            val args = IndexingRequestArgs()
                .also { it.parseArgs(workRequest.argumentsList.toTypedArray()) }

            val name = args.name ?: throw IllegalArgumentException("No name")
            val target = args.target ?: throw IllegalArgumentException("No target")
            val outPath = Paths.get(args.outDir ?: throw IllegalArgumentException("No output path"))

            val request = IndexRequest.newBuilder()
                .setProjectId(projectId)
                .setIndexDebugName(target)
                .setIndexId(
                    target
                        .replace("//", "")
                        .replace("/", "_")
                        .replace(":", "_")
                )
                .addAllSources(args.sources)
                .setProjectRoot(System.getProperty("user.dir"))
                .build()

            val response = client.index(request).getOrThrow()

            val moveOrThrow = { input: String, postfix: String ->
                val output = outPath.resolve(name + postfix).toFile()
                if (!File(input).renameTo(output)) {
                    throw IOException("Can't move $input to $output")
                }
            }

            moveOrThrow(response.ijxPath, ".ijx")
            moveOrThrow(response.ijxMetadataPath, ".ijx.metadata.json")
            moveOrThrow(response.ijxSha256Path, ".ijx.sha256")

            return WorkResponse.newBuilder()
                .setRequestId(workRequest.requestId)
                .build()
        } catch (e: Exception) {
            return WorkResponse.newBuilder()
                .setExitCode(1)
                .setOutput(e.toString())
                .setRequestId(workRequest.requestId)
                .build()
        }
    }
}